packages/wallet/src/default-rollup-block-submitter.ts

```diff
@@ -0,0 +1,230 @@
+  }
+
+  public async submitBlock(rollupBlock: RollupBlock): Promise<void> {
+    if (rollupBlock.number <= this.lastQueued) {
+      log.error(
+        `submitBlock(...) called on old block. Last Queued: ${
+          this.lastQueued
+        }, received: ${JSON.stringify(rollupBlock)}`
+      )
+      return
+    }
+
+    this.blockQueue.push(rollupBlock)
+    await this.db.put(
+      DefaultRollupBlockSubmitter.getBlockKey(rollupBlock.number),
+      DefaultRollupBlockSubmitter.serializeRollupBlockForStorage(rollupBlock)
+    )
+
+    this.lastQueued = rollupBlock.number
+    await this.db.put(
+      DefaultRollupBlockSubmitter.LAST_QUEUED_KEY,
+      this.numberToBuffer(this.lastQueued)
+    )
+
+    await this.trySubmitNextBlock()
+  }
+
+  public async handleNewRollupBlock(rollupBlockNumber: number): Promise<void> {
+    if (!this.blockQueue.length) {
+      log.error(
+        `Received block when no blocks pending. Block #: ${JSON.stringify(
+          rollupBlockNumber
+        )}`
+      )
+      return
+    }
+
+    if (rollupBlockNumber === this.blockQueue[0].number) {
+      this.blockQueue.shift()
+      this.lastConfirmed = rollupBlockNumber
+      await this.db.put(
+        DefaultRollupBlockSubmitter.LAST_CONFIRMED_KEY,
+        this.numberToBuffer(this.lastConfirmed)
+      )
+
+      // If we failed after submission but before storing submitted, update lastSubmitted
+      if (this.lastSubmitted < this.lastConfirmed) {
+        this.lastSubmitted = rollupBlockNumber
+        await this.db.put(
+          DefaultRollupBlockSubmitter.LAST_SUBMITTED_KEY,
+          this.numberToBuffer(this.lastSubmitted)
+        )
+      }
+      await this.trySubmitNextBlock()
+    }
+  }
+
+  private async trySubmitNextBlock() {
+    // If block has been submitted and is pending, return
+    if (
+      this.lastSubmitted > this.lastConfirmed ||
+      this.lastSubmitted >= this.lastQueued ||
+      !this.blockQueue.length
+    ) {
+      log.debug(
+        `Next block queued but not submitted because block ${this.lastSubmitted} was submitted but not yet confirmed.`
+      )
+      return
+    }
+
+    const block: RollupBlock = this.blockQueue.shift()
+
+    log.debug(
+      `Submitting block number ${block.number}: ${JSON.stringify(block)}.`
+    )
+
+    try {
+      const receipt = await this.rollupContract.submitBlock(
+        DefaultRollupBlockSubmitter.serializeRollupBlockForSubmission(block)
+      )
+      // TODO: do something with receipt?
+    } catch (e) {
+      logError(
+        log,
+        `Error submitting rollup block: ${JSON.stringify(block)}`,
+        e
+      )
+      throw e
+    }
+
+    this.lastSubmitted = block.number
+    await this.db.put(
+      DefaultRollupBlockSubmitter.LAST_SUBMITTED_KEY,
+      this.numberToBuffer(this.lastSubmitted)
+    )
+  }
+
+  public static serializeRollupBlockForSubmission(
+    block: RollupBlock
+  ): string[] {
+    return block.transitions.map((x) => x.stateRoot)
```

Comment:
Hmmmm, seems that Jing's wifi squashed a comment I had left here or at least I can't find it now--only change I was actually requesting with this original review was that this should be `(x) => abiEncodeTransition(x)`; contract needs the full transition to be rolled up.

Reply:
Good call on this! Updated 👍